File: ponder_dblp.pl
# TODO
# if authors re-occur, do we want to add the ORCID?

use IO::Uncompress::Gunzip qw(gunzip $GunzipError);
use strict;
use warnings;

# open dblp.rdf
my $dblp_file = 'dblp.rdf.gz';
# open my $fh, '<:gzip', $dblp_file or die "Could not open file '$dblp_file' $!";
my $fh = IO::Uncompress::Gunzip->new($dblp_file) or die "IO::Uncompress::Gunzip failed: $GunzipError\n";

open my $papers_fh, '>', 'dblp_papers.csv' or die "Could not open file 'dblp_papers.csv' $!";
print $papers_fh "NumericID\tDBLP\tTitle\tYear\n";

open my $authors_fh, '>', 'dblp_authors.csv' or die "Could not open file 'dblp_authors.csv' $!";
print $authors_fh "NumericID\tDBLP\tName\tORCID\n";

open my $papers_authors_fh, '>', 'dblp_papers_authors.csv' or die "Could not open file 'dblp_papers_authors.csv' $!";
print $papers_authors_fh "PaperID\tAuthorID\n";

# remember start time
my $start_time = time();
print "Start time: " . localtime($start_time) . "\n";

# states:
# 0 = searching
# 1 = inproceedings
# 2 = person
# 3 = article
# 4 = informal
# 5 = repository
# 6 = reference

my $state = 0;
my $currPaperID = '';
my $currTitle = '';
my $currName = '';

my $printLevel = 1; # 0 = no print, 1 = print error, 2 = print warning, 3 = print info

# hash for paper IDs
my %papersToNumbers = ();
my $maxPaperNumber = 1;
my $currPapernumber = 0;
# hash for author IDs
my %authorsToNumbers = ();
my $maxAuthorNumber = 1;
my $currAuthorNumber = 0;

my @authorsAndPapers = ();

my @currAuthors = ();
my @currCreators = ();
my $currYear = 0;

sub printInfo {
    my ($msg) = @_;
    if ($printLevel >= 3) {
        print "INFO: $msg\n";
    }
}

sub printWarning {
    my ($msg) = @_;
    if ($printLevel >= 2) {
        print "WARNING: $msg\n";
    }
}

sub printError {
    my ($msg) = @_;
    if ($printLevel >= 1) {
        print "ERROR: $msg\n";
    }
}

sub addPaper {
    my ($line) = @_;
    @currAuthors = ();
    @currCreators = ();

    $line =~ /rdf:about="([^"]+)"/;
    if ($1) {
        $currPaperID = $1;
        printInfo "Current ID: $currPaperID\n";
        $papersToNumbers{$currPaperID} = $maxPaperNumber;
        $currPapernumber = $maxPaperNumber;
        $maxPaperNumber++;
        printInfo "Assigned number $papersToNumbers{$currPaperID} to paper ID $currPaperID\n";
    } else {
        printError "No ID found in entry: $line\n";
    }
}

sub extractResource {
    my ($line) = @_;
    if ($line =~ /rdf:resource="([^"]+)"/) {
        return $1;
    } else {
        printError "No resource found in line: $line\n";
        return undef;
    }
}

sub checkAuthor {
    my ($currAuthorID, $currAuthorOrcid, $currAuthorName) = @_;
    my $realID = 0;
    if (exists $authorsToNumbers{$currAuthorID}) {
        printInfo "Found author ID: $currAuthorID\n";
        $realID = $authorsToNumbers{$currAuthorID};
    } else {
        printInfo "New author ID: $currAuthorID\n";
        $realID = $maxAuthorNumber;
        $authorsToNumbers{$currAuthorID} = $realID;
        printInfo "Assigned number $realID to author ID $currAuthorID\n";
        print $authors_fh "$realID\t$currAuthorID\t$currAuthorName\t$currAuthorOrcid\n";
        $maxAuthorNumber++;
    }
    return $realID;
}

sub processPaperContent {
    my ($line) = @_;
    my $currAuthorID = '';
    my $currAuthorOrcid = '';
    my $currAuthorName = '';

    if ($line =~ /<dblp:title>([^<]+)<\/dblp:title>/) {
        $currTitle = $1;
        printInfo "Current Title: $currTitle\n";
    } elsif (index($line, "<dblp:yearOfPublication") != -1) {
        if ($line =~ /<dblp:yearOfPublication.*?>(\d+)<\/dblp:yearOfPublication>/) {
            $currYear = $1;
            printInfo "Current Year: $currYear\n";
        } else {
            printError "No year found in line: $line\n";
        }
    } elsif ($line =~ /<dblp:authoredBy rdf:resource="([^"]+)"/) {
        printInfo "Found author ID: $1\n";
        push @currAuthors, $1;
    } elsif (index($line, "<dblp:AuthorSignature") != -1) {
        do {
            $line = <$fh>;
            chomp $line;
            if (index($line, "<dblp:signatureCreator") != -1) {
                if(my $res = extractResource($line)) {
                    $currAuthorID = $res;
                } else {
                    printError "No ID found in signatureCreator entry\n";
                }
            } elsif (index($line, "<dblp:signatureOrcid") != -1) {
                if(my $res = extractResource($line)) {
                    $currAuthorOrcid = $res;
                } else {
                    printError "No ID found in signatureOrcid entry\n";
                }
            } elsif (index($line, "<dblp:signatureDblpName") != -1) {
                if ($line =~ /<dblp:signatureDblpName>([^<]+)<\/dblp:signatureDblpName>/) {
                    $currAuthorName = $1;
                }
            }
        } while (index($line, "<\/dblp:AuthorSignature") == -1);
        push @currCreators, [$currAuthorID, $currAuthorOrcid, $currAuthorName];
        printInfo "Found creator: $currAuthorID, $currAuthorOrcid, $currAuthorName\n";
    }
}

sub finalizePaper {
    if ((scalar @currAuthors) != (scalar @currCreators)) {
        printError "Number of authors and creators do not match for paper $currPaperID: " . scalar(@currAuthors) . " vs " . scalar(@currCreators) . "\n";
    } else {
        foreach my $entry (@currCreators) {
            my ($currAuthorID, $currAuthorOrcid, $currAuthorName) = @$entry;
            my $realID = checkAuthor($currAuthorID, $currAuthorOrcid, $currAuthorName);
            print $papers_authors_fh "$currPapernumber\t$realID\n";
        }
    }
    print $papers_fh "$currPapernumber\t$currPaperID\t$currTitle\t$currYear\n";
}

while (my $line = <$fh>) {
    chomp $line;

    # switch depending on state
    if ($state == 0) {
        # Searching for the start of a new entry
        if (index($line, "<dblp:Inproceedings") != -1) {
            $state = 1;
            printInfo "Found Inproceedings entry\n";
            addPaper($line);
        } elsif (index($line, "<dblp:Article") != -1) {
            $state = 3;
            printInfo "Found Article entry\n";
            addPaper($line);
        }
        # elsif (index($line, "<dblp:Person") != -1) {
        #     $state = 2;
        #     printInfo "Found Person entry\n";
        #     $line =~ /rdf:about="([^"]+)"/;
        #     if ($1) {
        #         $currID = $1;
        #         printInfo "Current ID: $currID\n";
        #         $authorsToNumbers{$currID} = $maxAuthorNumber;
        #         $currAuthorNumber = $maxAuthorNumber;
        #         $maxAuthorNumber++;
        #         printInfo "Assigned number $authorsToNumbers{$currID} to author ID $currID\n";
        #     } else {
        #         printError "No ID found in Person entry\n";
        #     }
        # }

        # elsif ($line =~ /<dblp:informal/) {
        #     $state = 4;
        #     print "Found Informal entry\n";
        # } elsif ($line =~ /<dblp:repository/) {
        #     $state = 5;
        #     print "Found Repository entry\n";
        # } elsif ($line =~ /<dblp:reference/) {
        #     $state = 6;
        #     print "Found Reference entry\n";
        # }
    }

    if ($state == 1) {
        if (index($line, "<\/dblp:Inproceedings") != -1) {
            $state = 0;
            printInfo "End of Inproceedings entry\n";
            finalizePaper();
        } else {
            processPaperContent($line);
        }
    }
    if ($state == 3) {
        if (index($line, "<\/dblp:Article") != -1) {
            $state = 0;
            printInfo "End of Article entry\n";
            finalizePaper();
        } else {
            processPaperContent($line);
        }
    }
    # if ($state == 2) {
    #     if (index($line, "<\/dblp:Person") != -1) {
    #         $state = 0;
    #         printInfo "End of Person entry\n";
    #         print $authors_fh "$currAuthorNumber\t$currID\t$currName\n";
    #     } elsif ($line =~ /^<dblp:primaryCreatorName>([^<]+)<\/dblp:primaryCreatorName>/) {
    #         $currName = $1;
    #         printInfo "Current Author Name: $currName\n";
    #     }
    # }
}

# now clean up the author-paper relations
# foreach my $entry (@authorsAndPapers) {
#     my ($paperID, $authorID) = @$entry;
#     if (exists $authorsToNumbers{$authorID}) {
#         print $papers_authors_fh "$paperID\t$authorsToNumbers{$authorID}\n";
#     } else {
#         printError "Author ID not found in hash: $authorID\n";
#     }
# }

close $fh;
close $papers_fh;
close $authors_fh;
close $papers_authors_fh;

# get end time
my $end_time = time();
print "End time: " . localtime($end_time) . "\n";

# print elapsed time
my $elapsed_time = $end_time - $start_time;
my $hours = int($elapsed_time / 3600);
my $minutes = int(($elapsed_time % 3600) / 60);
my $seconds = $elapsed_time % 60;
print "Elapsed time: $hours hours, $minutes minutes, $seconds seconds\n";

File: ponder_dblp/ponder_dblp.cpp
#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <array>
#include <tuple>
#include <regex>
#include <filesystem>

// TODO
// profile the code
// is it better collect everything per thread and merge afterward? ID merging would be a headache
#define DEBUGGING 1
#define USE_THREAD_POOL 1

#define WITH_GZFILEOP
#include <zlib-ng.h>

#include <pugixml.hpp>

#include "InMemDB.hpp"
#include "ThreadPool.hpp"
#include "ThreadSafeIDGenerator.hpp"
#include "Timer.hpp"

// Utility functions for logging
enum class LogLevel { None, Error, Warning, Info };
LogLevel printLevel = LogLevel::Warning;

void printInfo(const std::string& msg) {
	if (printLevel >= LogLevel::Info) {
		std::cout << "INFO: " << msg << std::endl;
	}
}

void printWarning(const std::string& msg) {
	if (printLevel >= LogLevel::Warning) {
		std::cout << "WARNING: " << msg << std::endl;
	}
}

void printError(const std::string& msg) {
	if (printLevel >= LogLevel::Error) {
		std::cerr << "ERROR: " << msg << std::endl;
	}
}

ThreadSafeIDGenerator<uint32_t> authorsToNumbers;
struct Author {
	std::string id;
	std::string orcid;
	std::string name;
};
InMemDB<uint32_t, Author> authorDB;

ThreadSafeIDGenerator<uint32_t> papersToNumbers;
struct Paper {
	std::string id;
	std::string title;
	uint8_t type = 0;
	uint16_t year = 0;
};
InMemDB<uint32_t, Paper> paperDB;

LinkDB<uint32_t> papersAndAuthorsDB;

class ParserState {
public:
	enum Value : uint8_t {
		Searching = 0,
		Inproceedings,
		Incollection,
		Article,
		Book,
		Part,
		Informal,
		Data,
		NUMBER_OF_PARSER_STATES
	};

	inline static std::string startingPrefix = "<dblp:";
	inline static std::string endingPrefix = "</dblp:";

	constexpr operator Value() const {
		return state;
	}

	static std::string getEntityFromState(Value st) {
		switch (st) {
		case Inproceedings:
			return "Inproceedings";
		case Incollection:
			return "Incollection";
		case Article:
			return "Article";
		case Book:
			return "Book";
		case Part:
			return "Part";
		case Informal:
			return "Informal";
		case Data:
			return "Data";
		default:
			return "Unknown";
		}
	}

	std::string getEntity() const {
		return getEntityFromState(state);
	}

	static std::string getStartSnippet(Value st) {
		return startingPrefix + getEntityFromState(st);
	}

	static std::string getEndSnippet(Value st) {
		return endingPrefix + getEntityFromState(st) + ">";
	}

	std::string getStartSnippet() const {
		return getStartSnippet(state);
	}
	std::string getEndSnippet() const {
		return getEndSnippet(state);
	}

	void checkForStateChange(const std::string& line) {
		if (state == Searching) {
			if (line.find(startingPrefix) == std::string::npos) {
				return;
			}
			for (int i = 1; i < NUMBER_OF_PARSER_STATES; ++i) {
				if (line.find(getEntityFromState(static_cast<Value>(i)), startingPrefix.size()) == startingPrefix.size()) {
					state = static_cast<Value>(i);
					return;
				}
			}
		} else {
			if (line.find(endingPrefix) == std::string::npos) {
				return;
			}
			if (line.find(getEntity(), endingPrefix.size()) == endingPrefix.size()) {
				state = Searching;
			}
		}
	}

private:
	Value state = Searching;
};

// Helper functions
int checkAuthor(const std::string& authorID, const std::string& authorOrcid, const std::string& authorName) {
	auto res = authorsToNumbers.getOrCreateID(authorID);
	int realID = std::get<0>(res);
	if (std::get<1>(res)) {
		// New author, add to the database
		authorDB.storeItem(realID, { authorID, authorOrcid, authorName });
		printInfo("Assigned number " + std::to_string(realID) + " to author ID " + authorID);
	} else {
		// Existing author
		printInfo("Existing author found: " + authorID);
	}
	return realID;
}

std::string dump_node(pugi::xml_node n) {
	std::ostringstream oss;
	n.print(oss);
	return oss.str();
}

pugi::xml_node check_child(pugi::xml_node n, std::string name) {
	auto c = n.child(name);
	if (c == nullptr) throw std::invalid_argument("could not find '" + name + "' in '" + n.name() + "'\n" + dump_node(n));
	return c;
};

std::string check_attribute(pugi::xml_node n, std::string name) {
	auto r = n.attribute(name);
	if (r == nullptr) return "";
	return r.value();
};

std::string check_resource(pugi::xml_node n) {
	return check_attribute(n, "rdf:resource");
};

int processPaperBuffer(std::vector<std::string> buf, ParserState::Value paperType) {
	static std::regex idRegex(R"q(rdf:about="([^"]+)")q");
	static std::regex titleRegex(R"q(<dblp:title>([^<]+)</dblp:title>)q");
	static std::regex yearRegex(R"q(<dblp:yearOfPublication.*?>(\d+)</dblp:yearOfPublication>)q");
	static std::regex authorRegex(R"q(<dblp:authoredBy rdf:resource="([^"]+)")q");
	static std::regex signatureCreatorRegex(R"q(<dblp:signatureCreator rdf:resource="([^"]+)")q");
	static std::regex signatureOrcidRegex(R"q(<dblp:signatureOrcid rdf:resource="([^"]+)")q");
	static std::regex signatureNameRegex(R"q(<dblp:signatureDblpName>([^<]+)</dblp:signatureDblpName>)q");

	std::smatch match;

	std::vector<std::tuple<std::string, std::string, std::string>> currCreators;
	std::vector<std::string> currAuthors;
	std::string currPaperID;
	uint32_t currPaperNumericID = 0;
	std::string currTitle;
	uint16_t currYear = 0;
	std::string authorID, authorOrcid, authorName;

	std::ostringstream imploded;
	std::copy(buf.begin(), buf.end(),
		std::ostream_iterator<std::string>(imploded, "\n"));
	pugi::xml_document doc;
	auto result = doc.load_string(imploded.str().c_str());

	if (result.status != pugi::status_ok) {
		throw std::invalid_argument("could not parse publication: " + buf[0]);
	}

	auto pub = *doc.children().begin(); //doc.child("dblp:" + ParserState::getEntityFromState(paperType));
	currPaperID = check_attribute(pub, "rdf:about");
	if (currPaperID.empty()) {
		printWarning("No ID found, skipping:\n" + dump_node(pub));
		return 0;
	}
	printInfo("Current ID: " + currPaperID);
	auto res = papersToNumbers.getOrCreateID(currPaperID);
	currPaperNumericID = std::get<0>(res);
	printInfo("Assigned number " + std::to_string(currPaperNumericID) + " to paper ID " + currPaperID);

	for (pugi::xml_node author: pub.children("dblp:authoredBy")) {
		auto res = check_resource(author);
		currAuthors.push_back(res);
	}
	for (pugi::xml_node sig: pub.children("dblp:hasSignature")) {
		for (pugi::xml_node sig_content : sig.children("dblp:AuthorSignature")) {
			authorID = check_resource(check_child(sig_content, "dblp:signatureCreator"));
			try {
				authorOrcid = check_resource(check_child(sig_content, "dblp:signatureOrcid"));
			} catch (const std::invalid_argument&) {
				authorOrcid = "";
			}
			authorName = check_child(sig_content, "dblp:signatureDblpName").child_value();
			printInfo("Found creator: " + authorID + ", " + authorOrcid + ", " + authorName);
			currCreators.emplace_back(authorID, authorOrcid, authorName);
		}
	}

	if (currAuthors.size() != currCreators.size()) {
		printError("Number of authors and creators do not match for paper " + currPaperID + ": " +
			std::to_string(currAuthors.size()) + " vs " + std::to_string(currCreators.size()));
		return 0;
	}
	if (currAuthors.size() == 0) {
		printInfo("No authors found for paper " + currPaperID + ", skipping");
		// printWarning("\n" + dump_node(pub));
		return 0;
	}
	auto child_title = pub.child("dblp:title");
	if (child_title == nullptr) {
		printWarning("No title found for paper " + currPaperID + ", skipping\n" + dump_node(pub));
		return 0;
	}
	currTitle = child_title.child_value();
	printInfo("Current Title: " + currTitle);

	auto child_year = pub.child("dblp:yearOfPublication");
	if (child_year == nullptr) {
		child_year = pub.child("dblp:yearOfEvent");
		if (child_year == nullptr) {
			printWarning("No year found for paper " + currPaperID + ", skipping\n" + dump_node(pub));
			return 0;
		}
	}
	currYear = static_cast<uint16_t>(std::stoi(child_year.child_value()));
	printInfo("Current Year: " + std::to_string(currYear));

	for (const auto& [authorID, authorOrcid, authorName] : currCreators) {
		int realID = checkAuthor(authorID, authorOrcid, authorName);
		papersAndAuthorsDB.storeLink(currPaperNumericID, realID);
	}
	paperDB.storeItem(currPaperNumericID, { currPaperID, currTitle, paperType, currYear });

	return 0;
}

void checkProgress(uint64_t current, uint64_t total) {
	static uint64_t lastProgress = 0;
	const int barWidth = 70;

	if (current - lastProgress > 1) {
		auto progress = static_cast<double>(current) / total;
		std::cout << "[";
		int pos = barWidth * progress;
		for (int i = 0; i < barWidth; ++i) {
			if (i < pos) std::cout << "=";
			else if (i == pos) std::cout << ">";
			else std::cout << " ";
		}
		std::cout << "] " << std::setfill(' ') << std::setw(3) << int(progress * 100.0) << " %\r";
		//std::cout.flush();

		lastProgress = current;
	}
}

void dumpData() {
	// File streams
	std::ofstream papersFile("dblp_papers.csv");
	std::ofstream authorsFile("dblp_authors.csv");
	std::ofstream papersAuthorsFile("dblp_papers_authors.csv");

	// Initialize files
	papersFile << "NumericID\tDBLP\tTitle\tType\tYear\n";
	authorsFile << "NumericID\tDBLP\tName\tORCID\n";
	papersAuthorsFile << "PaperID\tAuthorID\n";

	std::cout << "Dumping papers..." << std::endl;
	for (auto p = 1; p < papersToNumbers.getMaxID(); ++p) {
		if (p % 100000 == 0) {
			checkProgress(p, papersToNumbers.getMaxID());
		}
		auto paper = paperDB.getItem(p);
		papersFile << p << "\t" << paper.id << "\t" << paper.title << "\t" << unsigned(paper.type) << "\t" << paper.year << "\n";
	}
	std::cout << std::endl << "Dumping authors..." << std::endl;
	for (auto a = 1; a < authorsToNumbers.getMaxID(); ++a) {
		if (a % 100000 == 0) {
			checkProgress(a, authorsToNumbers.getMaxID());
		}
		auto author = authorDB.getItem(a);
		authorsFile << a << "\t" << author.id << "\t" << author.name << "\t" << author.orcid << "\n";
	}
	std::cout << std::endl << "Dumping relations..." << std::endl;
	for (auto r = 1; r < papersAndAuthorsDB.size(); ++r) {
		if (r % 100000 == 0) {
			checkProgress(r, papersAndAuthorsDB.size());
		}
		auto link = papersAndAuthorsDB.getItem(r);
		papersAuthorsFile << link.first << "\t" << link.second << "\n";
	}
	std::cout << std::endl << "Dumping completed." << std::endl;

	// Close files
	papersFile.close();
	authorsFile.close();
	papersAuthorsFile.close();
}

void checkGZProgress(uint64_t lineCount, gzFile file, uint64_t total) {
	if (lineCount % 1000000 == 0) {
		auto pos = zng_gztell(file);
		if (pos > -1) {
			checkProgress(static_cast<uint64_t>(pos), total);
		}
	}
}

bool checkLockFile(const std::string& lockFilePath) {
	if (std::filesystem::exists(lockFilePath)) {
		std::cerr << "Lock file exists. Another instance may be running. Exiting.";
		return false;
	}
	// Create lock file
	std::ofstream lockFile(lockFilePath);
	if (!lockFile) {
		std::cerr << "Failed to create lock file. Exiting.";
		return false;
	}
	lockFile << "Lock file for ponder_dblp. Remove this file only if you are SURE no other instance is running.\n";
	return true;
}

void removeLockFile(const std::string& lockFilePath) {
	if (std::filesystem::exists(lockFilePath)) {
		std::filesystem::remove(lockFilePath);
	}
}

int main() {
#ifdef DEBUGGING
	// hack for broken visual studio cwd
	std::filesystem::current_path("u:\\src\\quickDBLP");
#endif

	//const std::string inputFilePath = "mini.rdf.gz";
	const std::string inputFilePath = "dblp.rdf.gz";
	const std::string lockFilePath = inputFilePath + ".lock";

	try {
#ifndef DEBUGGING
		if (!checkLockFile(lockFilePath)) {
			return 1;
		}
#endif

		std::vector<char> hugebuf;
		hugebuf.resize(1024 * 1024 * 256);

		auto file = zng_gzopen(inputFilePath.c_str(), "rb");
#ifdef DEBUGGING
		uint64_t fileSizeGZ = 38980527145;
#else
		uint64_t fileSizeGZ = 0;
		{
			Timer timer("Checking database size...");
			if (!file) {
				std::cerr << "Failed to open file with zlib: " << inputFilePath << "\n";
				removeLockFile(lockFilePath);
				return 1;
			}
			int numRead = 0;
			// start timer

			while ((numRead = zng_gzread(file, hugebuf.data(), hugebuf.size())) > 0) {
				fileSizeGZ += numRead;
			}
			zng_gzclose(file);
			std::cout << "File size (zlib): " << fileSizeGZ << " bytes\n";
		}
#endif
		
		file = zng_gzopen(inputFilePath.c_str(), "rb");
		{
			Timer timer("Processing database dump...");
			// Process file
			ThreadPool threadPool(std::thread::hardware_concurrency() * 2);
			std::string line;
			uint64_t lineCount = 0;
			int state = 0; // 0 = searching, 1 = inproceedings, 3 = article
			ParserState parserState, oldState;
			std::vector<std::string> paperBuffer;

			while (zng_gzgets(file, hugebuf.data(), hugebuf.size()) != nullptr) {
				line = hugebuf.data();
				checkGZProgress(++lineCount, file, fileSizeGZ);
				parserState.checkForStateChange(line);
				if (parserState != oldState) {
					if (oldState != ParserState::Searching) {
						printInfo("End of " + oldState.getEntity() + " entry");
						paperBuffer.push_back(line);
						ParserState::Value paperType = oldState;
#ifdef USE_THREAD_POOL
						threadPool.enqueue([paperBuffer, paperType]() { processPaperBuffer(paperBuffer, paperType); });
#else
						processPaperBuffer(paperBuffer, paperType);
#endif
					} else {
						printInfo("Found " + parserState.getEntity() + " entry");
						paperBuffer.clear();
						paperBuffer.push_back(line);
					}
					oldState = parserState;
				} else {
					if (parserState != ParserState::Searching) {
						paperBuffer.push_back(line);
					}
				}
			}
			std::cout << std::endl << "Waiting for threads to finish parsing..." << std::endl;
			threadPool.waitForAll();
		}
		std::cout << "saving CSVs..." << std::endl;
		dumpData();
	} catch (const std::exception& e) {
		printError("Exception: " + std::string(e.what()));
		removeLockFile(lockFilePath);
		return 1;
	} catch (...) {
		printError("Unknown exception occurred.");
		removeLockFile(lockFilePath);
		return 1;
	}
	removeLockFile(lockFilePath);
	return 0;
}